# KNN model

## Loading data

Step before loading date: Using excel first combined all the data in one sheet, and the sorting by the "DATE".

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('/Users/yunlei/Desktop/MGMT 478/Combined dataset.csv')
data.head(20)

STATION                                         NAME  LATITUDE  \
0   USW00014835   LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
1   USW00014848                    SOUTH BEND AIRPORT, IN US  41.70722   
2   USC00120784        BLOOMINGTON INDIANA UNIVERSITY, IN US  39.17399   
3   USW00014827      FORT WAYNE INTERNATIONAL AIRPORT, IN US  40.97248   
4   USW00093819    INDIANAPOLIS INTERNATIONAL AIRPORT, IN US  39.72515   
5   USW00053866         SHELBYVILLE MUNICIPAL AIRPORT, IN US  39.58545   
6   USW00093810  CARBONDALE SOUTHERN ILLINOIS AIRPORT, IL US  37.78329   
7   USW00094892   CHICAGO WEST CHICAGO DUPAGE AIRPORT, IL US  41.89641   
8   USW00094846   CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96017   
9   USC00114355                  ILLINOIS CITY DAM 16, IL US  41.42550   
10  USC00116011                         NASHVILLE 1 E, IL US  38.34189   
11  USW00023174     LOS ANGELES INTERNATIONAL AIRPORT, CA US  33.93816   
12  USW00023234   SAN FRANCISCO INTERNATIONAL AIRPORT, CA US  37.61962   
13  USW00023188       SAN DIEGO INTERNATIONAL AIRPORT, CA US  32.73360   
14  USW00093225       SACRAMENTO METROPOLITAN AIRPORT, CA US  38.70069   
15  USW00093193         FRESNO YOSEMITE INTERNATIONAL, CA US  36.77999   
16  USW00014835   LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
17  USW00014848                    SOUTH BEND AIRPORT, IN US  41.70722   
18  USC00120784        BLOOMINGTON INDIANA UNIVERSITY, IN US  39.17399   
19  USW00014827      FORT WAYNE INTERNATIONAL AIRPORT, IN US  40.97248   

    LONGITUDE  ELEVATION     DATE  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  
0   -86.94739      181.7  2010-01   8.5  1.01   NaN  23.5  29.5  17.5  
1   -86.31628      235.4  2010-01   9.4  1.22  27.4    23    29  16.9  
2   -86.52076      253.0  2010-01   NaN  1.89   7.0  24.9  31.5  18.2  
3   -85.20636      243.0  2010-01  10.7  0.64   7.7    23  28.9  17.2  
4   -86.28160      241.3  2010-01  10.1  1.22   7.8  24.6    31  18.2  
5   -85.79982      244.4  2010-01   9.4  1.30   NaN    25  31.4  18.7  
6   -89.24533      122.1  2010-01   7.6  1.33   NaN  29.1  37.5  20.8  
7   -88.25119      228.3  2010-01   9.4  0.69   NaN  19.9  26.5  13.3  
8   -87.93164      204.8  2010-01   9.8  1.13   9.1  21.9  27.4  16.4  
9   -91.00940      167.6  2010-01   NaN  2.00   7.6   NaN  23.9   NaN  
10  -89.35926      156.4  2010-01   NaN  1.51   4.6  25.4  32.8  18.1  
11 -118.38660       29.7  2010-01   6.0  4.30   NaN  58.4  66.8    50  
12 -122.36562        3.2  2010-01   6.5  5.97   NaN  51.2  55.6  46.8  
13 -117.18310        4.6  2010-01   4.5  3.38   NaN  58.2  66.5  49.9  
14 -121.59479        7.5  2010-01   6.5  5.61   NaN  46.9  52.7    41  
15 -119.72016      101.9  2010-01   4.3  2.06   NaN  48.6  55.5  41.7  
16  -86.94739      181.7  2010-02   7.6  0.61   NaN    26  32.8  19.1  
17  -86.31628      235.4  2010-02   8.3  1.35  18.5  25.6  31.3    20  
18  -86.52076      253.0  2010-02   NaN  2.13  15.3  26.3  33.7  18.8  
19  -85.20636      243.0  2010-02   9.4  1.07  12.1  25.6  31.8  19.4

In [3]:
# Check for missing values
data.isnull().sum()

STATION         0
NAME            0
LATITUDE        0
LONGITUDE       0
ELEVATION       0
DATE            0
AWND          700
PRCP            8
SNOW         1713
TAVG           27
TMAX           10
TMIN           25
dtype: int64

## Cleaning data

In [4]:
from sklearn.impute import SimpleImputer

In [5]:
# Drop rows where 'PRCP' is missing
data_cleaned = data.dropna(subset=['PRCP'])

In [6]:
# Drop the 'SNOW' column
data_cleaned = data_cleaned.drop(['SNOW'], axis=1)

In [7]:
data_cleaned.head()

STATION                                        NAME  LATITUDE  \
0  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
1  USW00014848                   SOUTH BEND AIRPORT, IN US  41.70722   
2  USC00120784       BLOOMINGTON INDIANA UNIVERSITY, IN US  39.17399   
3  USW00014827     FORT WAYNE INTERNATIONAL AIRPORT, IN US  40.97248   
4  USW00093819   INDIANAPOLIS INTERNATIONAL AIRPORT, IN US  39.72515   

   LONGITUDE  ELEVATION     DATE  AWND  PRCP  TAVG  TMAX  TMIN  
0  -86.94739      181.7  2010-01   8.5  1.01  23.5  29.5  17.5  
1  -86.31628      235.4  2010-01   9.4  1.22    23    29  16.9  
2  -86.52076      253.0  2010-01   NaN  1.89  24.9  31.5  18.2  
3  -85.20636      243.0  2010-01  10.7  0.64    23  28.9  17.2  
4  -86.28160      241.3  2010-01  10.1  1.22  24.6    31  18.2

In [8]:
# Convert non-numeric to numeric
for column in ['AWND', 'TAVG', 'TMAX', 'TMIN']:
    data_cleaned[column] = pd.to_numeric(data_cleaned[column], errors='coerce')

In [9]:
imputer = SimpleImputer(strategy='median')
data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']] = imputer.fit_transform(data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']])

In [10]:
data_cleaned.head(20)

STATION                                         NAME  LATITUDE  \
0   USW00014835   LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
1   USW00014848                    SOUTH BEND AIRPORT, IN US  41.70722   
2   USC00120784        BLOOMINGTON INDIANA UNIVERSITY, IN US  39.17399   
3   USW00014827      FORT WAYNE INTERNATIONAL AIRPORT, IN US  40.97248   
4   USW00093819    INDIANAPOLIS INTERNATIONAL AIRPORT, IN US  39.72515   
5   USW00053866         SHELBYVILLE MUNICIPAL AIRPORT, IN US  39.58545   
6   USW00093810  CARBONDALE SOUTHERN ILLINOIS AIRPORT, IL US  37.78329   
7   USW00094892   CHICAGO WEST CHICAGO DUPAGE AIRPORT, IL US  41.89641   
8   USW00094846   CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96017   
9   USC00114355                  ILLINOIS CITY DAM 16, IL US  41.42550   
10  USC00116011                         NASHVILLE 1 E, IL US  38.34189   
11  USW00023174     LOS ANGELES INTERNATIONAL AIRPORT, CA US  33.93816   
12  USW00023234   SAN FRANCISCO INTERNATIONAL AIRPORT, CA US  37.61962   
13  USW00023188       SAN DIEGO INTERNATIONAL AIRPORT, CA US  32.73360   
14  USW00093225       SACRAMENTO METROPOLITAN AIRPORT, CA US  38.70069   
15  USW00093193         FRESNO YOSEMITE INTERNATIONAL, CA US  36.77999   
16  USW00014835   LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
17  USW00014848                    SOUTH BEND AIRPORT, IN US  41.70722   
18  USC00120784        BLOOMINGTON INDIANA UNIVERSITY, IN US  39.17399   
19  USW00014827      FORT WAYNE INTERNATIONAL AIRPORT, IN US  40.97248   

    LONGITUDE  ELEVATION     DATE  AWND  PRCP  TAVG  TMAX   TMIN  
0   -86.94739      181.7  2010-01   8.5  1.01  23.5  29.5  17.50  
1   -86.31628      235.4  2010-01   9.4  1.22  23.0  29.0  16.90  
2   -86.52076      253.0  2010-01   7.8  1.89  24.9  31.5  18.20  
3   -85.20636      243.0  2010-01  10.7  0.64  23.0  28.9  17.20  
4   -86.28160      241.3  2010-01  10.1  1.22  24.6  31.0  18.20  
5   -85.79982      244.4  2010-01   9.4  1.30  25.0  31.4  18.70  
6   -89.24533      122.1  2010-01   7.6  1.33  29.1  37.5  20.80  
7   -88.25119      228.3  2010-01   9.4  0.69  19.9  26.5  13.30  
8   -87.93164      204.8  2010-01   9.8  1.13  21.9  27.4  16.40  
9   -91.00940      167.6  2010-01   7.8  2.00  57.5  23.9  47.55  
10  -89.35926      156.4  2010-01   7.8  1.51  25.4  32.8  18.10  
11 -118.38660       29.7  2010-01   6.0  4.30  58.4  66.8  50.00  
12 -122.36562        3.2  2010-01   6.5  5.97  51.2  55.6  46.80  
13 -117.18310        4.6  2010-01   4.5  3.38  58.2  66.5  49.90  
14 -121.59479        7.5  2010-01   6.5  5.61  46.9  52.7  41.00  
15 -119.72016      101.9  2010-01   4.3  2.06  48.6  55.5  41.70  
16  -86.94739      181.7  2010-02   7.6  0.61  26.0  32.8  19.10  
17  -86.31628      235.4  2010-02   8.3  1.35  25.6  31.3  20.00  
18  -86.52076      253.0  2010-02   7.8  2.13  26.3  33.7  18.80  
19  -85.20636      243.0  2010-02   9.4  1.07  25.6  31.8  19.40

In [11]:
# Convert 'DATE' to datetime and extract year and month
data_cleaned['DATE'] = pd.to_datetime(data_cleaned['DATE'])
data_cleaned['YEAR'] = data_cleaned['DATE'].dt.year
data_cleaned['MONTH'] = data_cleaned['DATE'].dt.month

In [12]:
# Drop the original 'DATE', 'STATION', and 'NAME' columns as they are not needed for prediction
data_final = data_cleaned.drop(['DATE', 'STATION', 'NAME'], axis=1)

In [13]:
data_final.head()

LATITUDE  LONGITUDE  ELEVATION  AWND  PRCP  TAVG  TMAX  TMIN  YEAR  MONTH
0  40.41236  -86.94739      181.7   8.5  1.01  23.5  29.5  17.5  2010      1
1  41.70722  -86.31628      235.4   9.4  1.22  23.0  29.0  16.9  2010      1
2  39.17399  -86.52076      253.0   7.8  1.89  24.9  31.5  18.2  2010      1
3  40.97248  -85.20636      243.0  10.7  0.64  23.0  28.9  17.2  2010      1
4  39.72515  -86.28160      241.3  10.1  1.22  24.6  31.0  18.2  2010      1

## Data Partition

In [14]:
# Determine the last year in the dataset
last_year = data_final['YEAR'].max()
last_year

2024

In [15]:
train_data = data_final[data_final['YEAR'] < last_year]
test_data = data_final[data_final['YEAR'] == last_year]

In [16]:
# Separate features and target variable
X_train = train_data.drop('PRCP', axis=1)
y_train = train_data['PRCP']
X_test = test_data.drop('PRCP', axis=1)
y_test = test_data['PRCP']

In [17]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

In [18]:
# Initialize a scaler
scaler = StandardScaler()

In [19]:
# Fit on training data and transform both training and test data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [20]:
# Initialize KNN regressor with default k=5
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X_train_scaled, y_train)
y_train_pred = knn.predict(X_train_scaled)
y_test_pred = knn.predict(X_test_scaled)

In [21]:
mse_train = mean_squared_error(y_train, y_train_pred)
mse_train

1.8918587281733745

In [22]:
mse_test = mean_squared_error(y_test, y_test_pred)
mse_test

2.9243693333333334

In [23]:
prcp_min = data_final['PRCP'].min()
prcp_max = data_final['PRCP'].max()
prcp_range = prcp_max - prcp_min
prcp_min, prcp_max, prcp_range

(0.0, 13.5, 13.5)

The average squared difference between the predicted and actual precipitation values of the model is 2.924369.

The problem for this model is that it needs input data to predict.

## KNN model ( Using Lag precipitation)

In [24]:
data = pd.read_csv('/Users/yunlei/Desktop/MGMT 478/Combined dataset_nonsort.csv')
data.head(20)

STATION                                        NAME  LATITUDE  \
0   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
1   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
2   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
3   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
4   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
5   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
6   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
7   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
8   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
9   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
10  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
11  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
12  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
13  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
14  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
15  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
16  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
17  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
18  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
19  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   

    LONGITUDE  ELEVATION     DATE  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  
0   -86.94739      181.7  2010-01   8.5  1.01   NaN  23.5  29.5  17.5  
1   -86.94739      181.7  2010-02   7.6  0.61   NaN    26  32.8  19.1  
2   -86.94739      181.7  2010-03   7.2  3.22   NaN  44.8  55.1  34.6  
3   -86.94739      181.7  2010-04   8.1  2.49   NaN  58.1  70.4  45.8  
4   -86.94739      181.7  2010-05   6.7  5.55   NaN  64.7  75.2  54.2  
5   -86.94739      181.7  2010-06   5.6  9.61   NaN  74.6  84.3  64.8  
6   -86.94739      181.7  2010-07   4.3  4.04   NaN  76.8  86.8  66.9  
7   -86.94739      181.7  2010-08   3.8  2.23   NaN  76.8  87.7    66  
8   -86.94739      181.7  2010-09   6.0  1.81   NaN  67.5    80    55  
9   -86.94739      181.7  2010-10   6.3  0.97   NaN  56.4  70.4  42.4  
10  -86.94739      181.7  2010-11   7.4  3.50   NaN  43.6  55.2    32  
11  -86.94739      181.7  2010-12   7.4  1.26   NaN  24.1    31  17.2  
12  -86.94739      181.7  2011-01   7.6  0.73   NaN  22.7  30.1  15.3  
13  -86.94739      181.7  2011-02   8.5  2.66   NaN  31.3  39.6    23  
14  -86.94739      181.7  2011-03   9.2  2.15   NaN  42.3  52.1  32.5  
15  -86.94739      181.7  2011-04  10.1  7.89   NaN  54.5  65.4  43.7  
16  -86.94739      181.7  2011-05   6.7  1.32   NaN  63.5  73.4  53.7  
17  -86.94739      181.7  2011-06   6.3  6.51   NaN    73  82.7  63.4  
18  -86.94739      181.7  2011-07   4.3  3.41   NaN  79.9    90  69.8  
19  -86.94739      181.7  2011-08   4.3  4.97   NaN  74.4  85.6  63.1

In [25]:
data['DATE'] = pd.to_datetime(data['DATE'])
data.sort_values(['STATION', 'DATE'], inplace=True)

In [26]:
data.set_index(['STATION', 'DATE'], inplace=True)

In [27]:
for lag in [1, 3, 7]:
    data[f'PRCP_Lag{lag}'] = data.groupby(level='STATION')['PRCP'].shift(lag)

In [28]:
data.head(20)

NAME  LATITUDE  LONGITUDE  \
STATION     DATE                                                           
USC00114355 2010-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

                        ELEVATION  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  \
STATION     DATE                                                        
USC00114355 2010-01-01      167.6   NaN  2.00   7.6   NaN  23.9   NaN   
            2010-02-01      167.6   NaN  0.80  12.8   NaN  31.1   NaN   
            2010-03-01      167.6   NaN  2.50   1.4   NaN  52.8   NaN   
            2010-04-01      167.6   NaN  3.34   0.0   NaN  71.4   NaN   
            2010-05-01      167.6   NaN  6.63   0.0   NaN    73   NaN   
            2010-06-01      167.6   NaN  9.96   0.0   NaN  83.6   NaN   
            2010-08-01      167.6   NaN  5.50   0.0   NaN    86   NaN   
            2010-09-01      167.6   NaN  4.76   0.0   NaN  76.6   NaN   
            2010-10-01      167.6   NaN  1.67   0.0   NaN  69.9   NaN   
            2010-11-01      167.6   NaN  1.55   0.0   NaN  52.8   NaN   
            2010-12-01      167.6   NaN  1.55  11.0   NaN  30.4   NaN   
            2011-01-01      167.6   NaN  0.89   NaN   NaN  26.6   NaN   
            2011-02-01      167.6   NaN  2.67   NaN   NaN  35.4   NaN   
            2011-03-01      167.6   NaN  2.07   NaN   NaN  48.4   NaN   
            2011-04-01      167.6   NaN  3.64   NaN   NaN  58.7   NaN   
            2011-05-01      167.6   NaN  3.51   NaN   NaN  73.4   NaN   
            2011-06-01      167.6   NaN  4.71   NaN  71.7  80.9  62.6   
            2011-07-01      167.6   NaN  2.02   0.0  80.6  89.8  71.4   
            2011-08-01      167.6   NaN  0.81   NaN  75.1  85.1    65   
            2011-09-01      167.6   NaN  2.21   NaN  62.3  70.9  53.7   

                        PRCP_Lag1  PRCP_Lag3  PRCP_Lag7  
STATION     DATE                                         
USC00114355 2010-01-01        NaN        NaN        NaN  
            2010-02-01       2.00        NaN        NaN  
            2010-03-01       0.80        NaN        NaN  
            2010-04-01       2.50       2.00        NaN  
            2010-05-01       3.34       0.80        NaN  
            2010-06-01       6.63       2.50        NaN  
            2010-08-01       9.96       3.34        NaN  
            2010-09-01       5.50       6.63       2.00  
            2010-10-01       4.76       9.96       0.80  
            2010-11-01       1.67       5.50       2.50  
            2010-12-01       1.55       4.76       3.34  
            2

In [46]:
data.dropna(subset=['PRCP'] + [f'PRCP_Lag{lag}' for lag in [1, 3, 7]], inplace=True)

In [47]:
data.head(20)

STATION       DATE                         NAME  LATITUDE  LONGITUDE  \
0   USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
1   USC00114355 2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
2   USC00114355 2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
3   USC00114355 2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
4   USC00114355 2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
5   USC00114355 2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
6   USC00114355 2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
7   USC00114355 2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
8   USC00114355 2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
9   USC00114355 2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
10  USC00114355 2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
11  USC00114355 2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
12  USC00114355 2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
13  USC00114355 2011-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
14  USC00114355 2011-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
15  USC00114355 2011-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
16  USC00114355 2012-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
17  USC00114355 2012-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
18  USC00114355 2012-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
19  USC00114355 2012-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

    ELEVATION  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  PRCP_Lag1  PRCP_Lag3  \
0       167.6   NaN  4.76   0.0   NaN  76.6   NaN       5.50       6.63   
1       167.6   NaN  1.67   0.0   NaN  69.9   NaN       4.76       9.96   
2       167.6   NaN  1.55   0.0   NaN  52.8   NaN       1.67       5.50   
3       167.6   NaN  1.55  11.0   NaN  30.4   NaN       1.55       4.76   
4       167.6   NaN  0.89   NaN   NaN  26.6   NaN       1.55       1.67   
5       167.6   NaN  2.67   NaN   NaN  35.4   NaN       0.89       1.55   
6       167.6   NaN  2.07   NaN   NaN  48.4   NaN       2.67       1.55   
7       167.6   NaN  3.64   NaN   NaN  58.7   NaN       2.07       0.89   
8       167.6   NaN  3.51   NaN   NaN  73.4   NaN       3.64       2.67   
9       167.6   NaN  4.71   NaN  71.7  80.9  62.6       3.51       2.07   
10      167.6   NaN  2.02   0.0  80.6  89.8  71.4       4.71       3.64   
11      167.6   NaN  0.81   NaN  75.1  85.1    65       2.02       3.51   
12      167.6   NaN  2.21   NaN  62.3  70.9  53.7       0.81       4.71   
13      167.6   NaN  0.87   NaN    56  66.9  45.2       2.21       2.02   
14      167.6   NaN  2.59   NaN  43.1  51.7  34.6       0.87       0.81   
15      167.6   NaN  1.87   0.0  33.7    41  26.3       2.59       2.21   
16      167.6   NaN  0.62   6.9  28.7  37.9  19.5       1.87       0.87   
17      167.6   NaN  1.86   NaN  32.1  39.6  24.5       0.62       2.59   
18      167.6   NaN  1.50   NaN  52.8  63.3  42.3       1.86       1.87   
19      167.6   NaN  4.50   0.0  53.1  64.1  42.2       1.50       0.62   

    PRCP_Lag7  
0        2.00  
1        0.80  
2        2.50  
3        3.34  
4        6.63  
5        9.96  
6        5.50  
7        4.76  
8        1.67  
9        1.55  
10       1.55  
11       0.89  
12       2.67  
13       2.07  
14       3.64  
15       3.51  
16       4.71  
17       2.02  
18       0.81  
19       2.21

In [48]:
data.reset_index(inplace=True)

In [49]:
data.head()

index      STATION       DATE                         NAME  LATITUDE  \
0      0  USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   
1      1  USC00114355 2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   
2      2  USC00114355 2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   
3      3  USC00114355 2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   
4      4  USC00114355 2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   

   LONGITUDE  ELEVATION  AWND  PRCP  SNOW TAVG  TMAX TMIN  PRCP_Lag1  \
0   -91.0094      167.6   NaN  4.76   0.0  NaN  76.6  NaN       5.50   
1   -91.0094      167.6   NaN  1.67   0.0  NaN  69.9  NaN       4.76   
2   -91.0094      167.6   NaN  1.55   0.0  NaN  52.8  NaN       1.67   
3   -91.0094      167.6   NaN  1.55  11.0  NaN  30.4  NaN       1.55   
4   -91.0094      167.6   NaN  0.89   NaN  NaN  26.6  NaN       1.55   

   PRCP_Lag3  PRCP_Lag7  
0       6.63       2.00  
1       9.96       0.80  
2       5.50       2.50  
3       4.76       3.34  
4       1.67       6.63

In [50]:
# Check for missing values
data.isnull().sum()

index           0
STATION         0
DATE            0
NAME            0
LATITUDE        0
LONGITUDE       0
ELEVATION       0
AWND          649
PRCP            0
SNOW         1628
TAVG           16
TMAX            6
TMIN           14
PRCP_Lag1       0
PRCP_Lag3       0
PRCP_Lag7       0
dtype: int64

In [51]:
# Drop the 'SNOW' column
data_cleaned = data.drop(['SNOW'], axis=1)

In [52]:
# Convert non-numeric to numeric
for column in ['AWND', 'TAVG', 'TMAX', 'TMIN']:
    data_cleaned[column] = pd.to_numeric(data_cleaned[column], errors='coerce')

In [53]:
imputer = SimpleImputer(strategy='median')
data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']] = imputer.fit_transform(data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']])

In [54]:
data_cleaned.head(20)

index      STATION       DATE                         NAME  LATITUDE  \
0       0  USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   
1       1  USC00114355 2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   
2       2  USC00114355 2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   
3       3  USC00114355 2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   
4       4  USC00114355 2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   
5       5  USC00114355 2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   
6       6  USC00114355 2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   
7       7  USC00114355 2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   
8       8  USC00114355 2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   
9       9  USC00114355 2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   
10     10  USC00114355 2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   
11     11  USC00114355 2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   
12     12  USC00114355 2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   
13     13  USC00114355 2011-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   
14     14  USC00114355 2011-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   
15     15  USC00114355 2011-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   
16     16  USC00114355 2012-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   
17     17  USC00114355 2012-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   
18     18  USC00114355 2012-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   
19     19  USC00114355 2012-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   

    LONGITUDE  ELEVATION  AWND  PRCP  TAVG  TMAX  TMIN  PRCP_Lag1  PRCP_Lag3  \
0    -91.0094      167.6   7.8  4.76  57.5  76.6  47.6       5.50       6.63   
1    -91.0094      167.6   7.8  1.67  57.5  69.9  47.6       4.76       9.96   
2    -91.0094      167.6   7.8  1.55  57.5  52.8  47.6       1.67       5.50   
3    -91.0094      167.6   7.8  1.55  57.5  30.4  47.6       1.55       4.76   
4    -91.0094      167.6   7.8  0.89  57.5  26.6  47.6       1.55       1.67   
5    -91.0094      167.6   7.8  2.67  57.5  35.4  47.6       0.89       1.55   
6    -91.0094      167.6   7.8  2.07  57.5  48.4  47.6       2.67       1.55   
7    -91.0094      167.6   7.8  3.64  57.5  58.7  47.6       2.07       0.89   
8    -91.0094      167.6   7.8  3.51  57.5  73.4  47.6       3.64       2.67   
9    -91.0094      167.6   7.8  4.71  71.7  80.9  62.6       3.51       2.07   
10   -91.0094      167.6   7.8  2.02  80.6  89.8  71.4       4.71       3.64   
11   -91.0094      167.6   7.8  0.81  75.1  85.1  65.0       2.02       3.51   
12   -91.0094      167.6   7.8  2.21  62.3  70.9  53.7       0.81       4.71   
13   -91.0094      167.6   7.8  0.87  56.0  66.9  45.2       2.21       2.02   
14   -91.0094      167.6   7.8  2.59  43.1  51.7  34.6       0.87       0.81   
15   -91.0094      167.6   7.8  1.87  33.7  41.0  26.3       2.59       2.21   
16   -91.0094      167.6   7.8  0.62  28.7  37.9  19.5       1.87       0.87   
17   -91.0094      167.6   7.8  1.86  32.1  39.6  24.5       0.62       2.59   
18   -91.0094      167.6   7.8  1.50  52.8  63.3  42.3       1.86       1.87   
19   -91.0094      167.6   7.8  4.50  53.1  64.1  42.2       1.50       0.62   

    PRCP_Lag7  
0        2.00  
1        0.80  
2        2.50  
3        3.34  
4        6.63  
5        9.96  
6        5.50  
7        4.76  
8        1.67  
9        1.55  
10       1.55  
11       0.89  
12       2.67  
13       2.07  
14       3.64  
15       3.51  
16       4.71  
17       2.02  
18       0.81  
19       2.21

In [55]:
data_cleaned['YEAR'] = data_cleaned['DATE'].dt.year
train_data = data_cleaned[data_cleaned['YEAR'] < data_cleaned['YEAR'].max()]
test_data = data_cleaned[data_cleaned['YEAR'] == data_cleaned['YEAR'].max()]

In [56]:
X_train = train_data[[f'PRCP_Lag{lag}' for lag in [1, 3, 7]]]
X_train.head()

PRCP_Lag1  PRCP_Lag3  PRCP_Lag7
0       5.50       6.63       2.00
1       4.76       9.96       0.80
2       1.67       5.50       2.50
3       1.55       4.76       3.34
4       1.55       1.67       6.63

In [57]:
y_train = train_data['PRCP']
X_test = test_data[[f'PRCP_Lag{lag}' for lag in [1, 3, 7]]]
y_test = test_data['PRCP']

In [58]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test) 

In [59]:
y_train.isnull().any().any()

False

In [64]:
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X_train_scaled, y_train)
y_test_pred = knn.predict(X_test_scaled)
y_test_pred[:5]
y_train_pred = knn.predict(X_train_scaled)

In [65]:
mse_train = mean_squared_error(y_train, y_train_pred)
mse_train

2.845744840731071

In [63]:
mse_test = mean_squared_error(y_test, y_test_pred)
mse_test 

2.1179003333333335